Extensions/xkit_patches.js

```diff
@@ -269,6 +269,329 @@ XKit.extensions.xkit_patches = new Object({
+								tag = tag.substring(1);
+							}
+
+							if (!post.tags) {
+								post.tags = tag;
+							} else {
+								post.tags += "," + tag;
+							}
+						});
+					}
+
+					return post;
+				}
+			};
+
+			/**
+			 * Instantiate and add a previously "created" button to the
+			 * specified post.
+			 * @param {Object} obj - Interface Post Object
+			 * @param {String} class_name - CSS class of the button to be added
+			 * @param {String?} additional - String inserted into the button's div tag
+			 */
+			XKit.interface.add_control_button = function(obj, class_name, additional) {
+
+				if (typeof additional == "undefined") {additional = ""; }
+
+				if (XKit.interface.added_icon.indexOf(class_name) === -1) {
+					// console.log("Interface -> Can't add icon, button not created, use create_control_button.");
+					return;
+				}
+
+				var m_text = XKit.interface.added_icon_text[XKit.interface.added_icon.indexOf(class_name)];
+
+				var post_obj = XKit.interface.parse_post(obj);
+				var post_id = post_obj.id;
+				var post_type = post_obj.type;
+				var post_permalink = post_obj.permalink;
+
+				var m_data = "data-post-id = \"" + post_id + "\" data-post-type=\"" + post_type + "\" data-permalink=\"" + post_permalink + "\"";
+
+				var m_html = "<div " + m_data + " title=\"" + m_text + "\" class=\"xkit-interface-control-button post_control post_control_icon " + class_name + "\" " + additional + "></div>";
+
+				if ($(obj).find(".post_controls_inner").length > 0) {
+					$(obj).find(".post_controls_inner").prepend(m_html);
+				} else if (XKit.browser().mobile) {
+					$(obj).find(".mh_post_foot_control.show_notes").after(m_html);
+				} else {
+					$(obj).find(".post_controls").prepend(m_html);
+				}
+
+				if (XKit.interface.where().search) {
+					XKit.interface.trigger_reflow();
+				}
+			};
+
+			/**
+			 * @param {String} post_id
+			 * @return {Object} Interface Post Object of post with given id
+			 */
+			XKit.interface.find_post = function(post_id) {
+
+				// Return a post object based on post ID.
+
+				if ($("body").find("#post_" + post_id).length > 0) {
+					return XKit.interface.parse_post($("#post_" + post_id));
+				} else if ($(".mh_post").length > 0) {
+					return XKit.interface.parse_post($(".mh_post"));
+				} else {
+					var m_error = {};
+					m_error.error = true;
+					m_error.error_message = "Object not found on page.";
+					return m_error;
+				}
+
+			};
+
+			XKit.interface.post = { /* globals Promise */
+
+				fetch: function(post) {
+					return new Promise(function(resolve, reject) {
+						XKit.tools.Nx_XHR({
+							method: "POST",
+							url: "https://www.tumblr.com/svc/post/fetch",
+							headers: {
+								"X-Tumblr-Form-Key": XKit.interface.form_key(),
+								"X-Requested-With": "XMLHttpRequest"
+							},
+							json: true, data: JSON.stringify(post),
+							onerror: function(response) {
+								reject(response);
+							},
+							onload: function(response) {
+								try {
```

Comment:
this whole function can be replaced with `resolve(response.reponseText)` and `.then(JSON.parse)`, avoiding the need for a try-catch and confusing control flow.